### ISSUE TITLE
Increase salt length (generic)

### ISSUE THREAD

**Issue by jsteube** [hashcat/hashcat#565]
There have been several issues for this topic and I wanted to merge them so we have a single issue for discussion this.

So when I think of increasing salt length I think of this:

* How much length is enough. It can't be unlimited (because GPU memory can't be dynamically allocated inside the kernel)
* Do we really want to make this generic for all kernels or just the "generic" kernels like MD5($p.$s). It's not always usefull. For example, phpass always uses a 8 byte salt. 
* We'd need to rewrite "generic" kernels and add loops (much slower), plus the merkle demgard handling, but I wouldn't count this as a "loss" in speed
* We'd need to rewrite many of the common.cl and host functions, propably removing manually unrolled code into (much slower) loops

So basically this is all about speed and it basically goes down to this one question I think:

* Can we accept the performance loss, or, how much loss in percentage is acceptable?

Any other ideas?

**Comment by jsteube** [comment]
Some more discussion material:

From @magnumripper 

> JtR core supports arbitrarily long salts but most optimized formats has a fairly tight limit just like Hashcat, to keep things in one or two Merkel Damgard blocks.
> 
> This particular format is supported in JtR as dynamic_2009 (only CPU) with salt lengths up to 200 (same format with max. 23 bytes salt is dynamic_9 and it's obviously faster).

From https://github.com/hashcat/hashcat/issues/79

> A new limit on length to 35-40 on the most common algorithms

From https://github.com/hashcat/hashcat/issues/95

> SaltLen = 130

From https://github.com/hashcat/hashcat/issues/442

> ... whatever hashcat supports salt length 40 or not.

**Comment by magnumripper** [comment]
Does your core code allow launching different kernels for different salts? If so, you could have a kernel optimized for a single block, another for two blocks and a third (slower) for arbitrary length.

If this isn't possible with current code, perhaps it's not too hard to implement?

**Comment by jsteube** [comment]
There's such a system for the different passwords lengths (to make automatic use of zero-based optimizations). I think we can do the same for salt lengths.

**Comment by Ins1ghtLabs** [comment]
+1. 
I got an algorithm that is sha1(unicode($pass.$salt)) and the salt is a fixed 40 char sha1 hash.

**Comment by polarathene** [comment]
Probably not an option, but is JIT compiling kernels perhaps with a template language if you want to have a variable salt that becomes static during compilation at runtime an option? ArrayFire does JIT compiling for it's kernels from memory.

**Comment by killswitch-GUI** [comment]
@jsteube Is this a issue with -m 7100 OSX v10.8+ as well? Interesting enough I'm experiencing issues where dict will work but brute force will not.

reference:
- salt 70 chars

output:
alexanders-MBP:loader alexanderrymdeko-harvey$ hashcat -a 3 -m 0 test.txt ?a?a?a?a?a?a
hashcat (v3.10) starting...

```
WARNING: Hashfile 'test.txt' on line 1 ($ml$41076$935aae561d7f865eb5e6f8fc4392c154630XXXXXXXd7bb24ef9a7a166d8bc4c10$c8ae662a491d2055ef304da8eed716dXXXXXXXXXXXXXd330f28cb72d93171d63ff613c0b5142c7bced73d5575388ec3843e71611bde2d32022dbb49cc09a8e1bf961a6c367d5b6d6fae0fa3e7f93ff4bf73c5245df97217b4962b399c1d92f7b60e81bf642f694ccdbede495ef49587005e7d332d7e5a3c): Line-length exception
Parsed Hashes: 2/2 (100.00%)
```

With dict: 
```
hashcat -a 0 -m 7100 test.txt dict.txt 
hashcat (v3.10) starting...

OpenCL Platform #1: Apple               
=========================
- Device #1: Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz, skipped
- Device #2: Iris Pro, 384/1536 MB allocatable, 40MCU
- Device #3: AMD Radeon R9 M370X Compute Engine, 512/2048 MB allocatable, 10MCU

Hashes: 1 hashes; 1 unique digests, 1 unique salts
Bitmaps: 16 bits, 65536 entries, 0x0000ffff mask, 262144 bytes, 5/13 rotates
Rules: 1
Applicable Optimizers:
* Zero-Byte
* Single-Hash
* Single-Salt
* Slow-Hash-SIMD
* Uses-64-Bit
Generated dictionary stats for dict.txt: 15 bytes, 1 words, 1 keyspace           

ATTENTION!                                                
  The wordlist or mask you are using is too small.
  Therefore, hashcat is unable to utilize the full parallelization power of your device(s).
  The cracking speed will drop.
  Workaround: https://hashcat.net/wiki/doku.php?id=frequently_asked_questions#how_to_create_more_work_for_full_speed

INFO: approaching final keyspace, workload adjusted       

[s]tatus [p]ause [r]esume [b]ypass [c]heckpoint [q]uit => s

Session.Name...: hashcat
Status.........: Running
Input.Mode.....: File (dict.txt)
Hash.Target....: $ml$48076$935aae5fc4392c15...
Hash.Type......: OSX v10.8+
Time.Started...: Thu Jan 12 11:42:14 2017 (8 secs)
Speed.Dev.#2...:        0 H/s (0.00ms)
Speed.Dev.#3...:        0 H/s (2.70ms)
Speed.Dev.#*...:        0 H/s
Recovered......: 0/1 (0.00%) Digests, 0/1 (0.00%) Salts
Progress.......: 0/1 (0.00%)
Rejected.......: 0/0 (0.00%)
Restore.Point..: 0/1 (0.00%)

$ml$42076$935aae561d7f865eb5e6f8fc4392cXXXXXX012df6ed7bb24ef9a7a166d8bc4c10$c8ae662a491d2055ef304da8eed7XXXXXXX85d28d8e234f6440721b7b33bfd330f28cb72d93171d63ff613c0b5142c7bced73d5575388ec3843e71XXXXde2d32:PASSWORD_AA
                                                          
Session.Name...: hashcat
Status.........: Cracked
Input.Mode.....: File (dict.txt)
Hash.Target....: $ml$48076$935aae5e6f8fc4392c15...
Hash.Type......: OSX v10.8+
Time.Started...: Thu Jan 12 11:42:14 2017 (14 secs)
Speed.Dev.#2...:        0 H/s (0.00ms)
Speed.Dev.#3...:        0 H/s (2.69ms)
Speed.Dev.#*...:        0 H/s
Recovered......: 1/1 (100.00%) Digests, 1/1 (100.00%) Salts
Progress.......: 1/1 (100.00%)
Rejected.......: 0/1 (0.00%)

Started: Thu Jan 12 11:42:14 2017
Stopped: Thu Jan 12 11:42:38 2017 
```

**Comment by philsmd** [comment]
@killswitch-GUI
This is clearly incorrect:
`hashcat -a 3 -m 0 test.txt ?a?a?a?a?a?a`

it should be something like this:
`hashcat -a 3 -m 7100 test.txt ?a?a?a?a?a?a`

The hash mode was incorrect (you set it to MD5 instead of OSX v10.8+) in your command line arguments

**Comment by chadgrant** [comment]
I have a hash dump with 64 chars of salt ... bummer hashcat can't do variable salt lengths since there are no restrictions on developers using very long salts

**Comment by jsteube** [comment]
I've made some experiments to find about what speed loss this will cause to the existing modes. Here's the results from my 750Ti using the same hash:salt combination for all 4 tests:

current sha1($salt.$pass):
-a0: 734.4 MH/s
-a3: 1047.2 MH/s

modified sha1($salt.$pass) with support for up to length 64 salt and length 64 password:
-a0: 665.5 MH/s (10% loss)
-a3: 864.5 MH/s (21% loss)

**Comment by gohashcat** [comment]
Hi! Its look nice to have speed loss, but can hashing more salt length.
I need to hash sha256(salt.pass) with pass+salt length more than 55 characters (its up to 80-90 characters long) But I can't crack it :( Or help, how can I crack it? 
Do you can to create algoritm sha256(salt.pass) or just sha256 (to use mask instead salt) version with longer salt+pass length (but slower hashing)?
Thanks! (sorry my English ;) )

**Comment by jsteube** [comment]
If you're interessted, I've pushed a branch which allows cracking of passwords and salts up to length 256 both here: https://github.com/hashcat/hashcat/tree/longer_passwords_and_salts

The drops are not that big as in previous attempts, but they are a bit more work intensive. Means each kernel would require a work intensive rewrite. For example, there would be a kernel that allows only combination of 16+16, 32+32 and all+all. This way the drop on MD5 (in -a 1 mode) is: 1451.5 MH/s -> 1374.6 MH/s. In -a 0 (rules) mode, the drop will be much bigger of course.

It is however only working with md5 and in -a 1 mode and it's more like an experiment. You can rewrite it to do sha256 if you have some programming experience.

**Comment by gituser** [comment]
@jsteube, I've tried that branch in both `-m 0` and `-m 2711` modes with `-a 1` and it hashcat didn't find any of the long (>128) passwords candidates. I can see in hashcat's status that such passwords are enumerated but not found for some reason.

`-m 2711` isn't supporting salts bigger than 30 chars.

I see that you've been adding commits to improve password length for other password hashes. 
So guess we just have to wait a bit.

Thanks a lot once again for your work, highly appreciated.

**Comment by jsteube** [comment]
Finished implementation with v4.0.0
